index.html

```diff
@@ -2479,12 +2483,12 @@ <h2>The <dfn>NFCAccept</dfn> enum</h2>
           </li>
           <li>
             Let <var>payload</var> be the result of invoking
-            <a>create a Web NFC Id</a> given <var>url</var>.
+            <a>create a Web NFC Id</a> given <var>urlPath</var>.
             If this throws an exception, re-throw it.
           </li>
           <li>
-            Set <var>ndef.PAYLOAD</var> to the <a>URL path</a> of the
-            <a>browsing context</a>, encoded as UTF-16.
+            Set <var>ndef.PAYLOAD</var> to the <a>URL path</a> of the current
+            <a>browsing context</a>, <a href="#dfn-encode">encoded</a> as <a>UTF-16 encoding</a>.
```

Comment:
Does NFC define some high-level way to set the PAYLOAD field from scalar values?

Which UTF-16 encoding is in use here? This would indeed require reverting https://github.com/whatwg/encoding/commit/8360f775c8df145f649047c7d59c5ff733ade112 somewhat. Perhaps an initial step would be to put those algorithms in NFC with an issue against whatwg/encoding for potential upstreaming.

Looking over this PR and skimming through Web NFC it seems like there's multiple places that are rather hand-wavy about how bytes and code points are handled. Those all need tidying up.

Reply:
@zolkis, could you comment on this?

Reply:
From the current spec and PR, it's very unclear what constraints lead to the notion that Web NFC needs to encode into UTF-16BE or UTF-16LE (which one?). At present, I think we aren't at the point where such a need is well enough motivated to change the Encoding Standard.

However, if there is a genuine need to enable taking a DOMString and generating bytes in UTF-16BE or UTF-16LE, it should probably be done outside the general framework of the Encoding Standard to make it absolutely clear that it doesn't apply to form submission, URL queries or the `TextEncoder` API.

Reply:
A few remarks based on the NDEF spec:
- NDEF mandates no specific processing model for NDEF messages. The usage of the payload types is entirely at the discretion of the user application.
- Encoding of URI characters which fall outside the US-ASCII range is left to the NDEF
application. Therefore, an NDEF parser must not assume any particular encoding for this field.
- To indicate the payload type, the TYPE field (max 255 bytes) defines it with the semantics set by the TNF field (e.g. URI, IANA media type, NFC Forum External Type, unknown, NFC Forum Well Known Type etc.).
- Transmission order is big-endian (most significant octet is transmitted first, and leftmost bit is the most significant bit).
- I don't remember any more why we specified UTF-16 at that time. WFIW `neard` supported UTF-8 and UTF-16 at that time. The encoding support might be limited by the implementation and otherwise should be handled by the application (from NFC pov the implementation is an application, so it can well define the required encoding; hence the Web NFC spec could also use/mandate a given encoding if there was a use case for that). For Web NFC we should comply with/refer to the (Web) Encoding Standard.

Reply:
(BTW apologies for the late answer, jut got back from holidays).

Reply:
If you could always produce UTF-8 and consume UTF-16(LE/BE?) as needed, that'd be ideal. It sounds like that might be doable?

(It might be good for someone to split this into a couple of issues as tracking all the various things in one PR will make it easy to overlook things.)

Reply:
Should it even be a string at all? It sounds like from your comments the application should probably be doing any encoding/decoding, and the Web NFC spec might want to just provide a mechanism for exchanging raw bytes.

Reply:
@domenic That is right, the spec should just provide a bitpipe to the app. For convenience we also have the freedom to pick a certain number of supported encodings but that would be a restriction from NFC point of view and would add complexity to the spec/implementations.

Reply:
@zolkis, so do we have a conclusion here? Should I create a separate issue to track encoding/decoding problem?

Reply:
Yes, we should track this in a separate issue. The conclusion is that we need mechanisms/algorithms for how to handle encoding. It's not a simple property.